util/src/journaldb.rs

```diff
@@ -402,15 +353,23 @@ impl HashDB for JournalDB {
 	}
 
 	fn lookup(&self, key: &H256) -> Option<&[u8]> { 
-		let k = self.overlay.raw(key);
+		let k = self.transaction_overlay.raw(key);
 		match k {
 			Some(&(ref d, rc)) if rc > 0 => Some(d),
 			_ => {
-				if let Some(x) = self.payload(key) {
-					Some(&self.overlay.denote(key, x).0)
-				}
-				else {
-					None
+				let v = self.journal_overlay.as_ref().map_or(None, |ref j| j.read().unwrap().backing_overlay.lookup(key).map(|v| v.to_vec()));
+				match v {
+					Some(x) => {
+						Some(&self.transaction_overlay.denote(key, x).0)
+					}
+					_ => {
+						if let Some(x) = self.payload(key) {
+							Some(&self.transaction_overlay.denote(key, x).0)
+						}
+						else {
+							None
+						}
+					}
```

Comment:
Tried this, got tests failing. I'm sure there is a functional way to express this, I'll think about it if this PR ever gets accepted.